### PR TITLE
[JIMT][SL-592] tutorial explorer dialog focus

### PR DESCRIPTION
**Pull request by thomasoniii**
## Summary

In the list of tutorials, when you pop open the tutorial detail dialog you would not be able to tab between items on the dialog, making it less accessible.

For starters, _technically_ you could tab on the dialog - the issue was that the dialog actually lived "early" in the page, so if you shift-tabbed a number of times until you fell off the start of the tutorial list, you'd then tab your way back up through the items on the modal until you eventually tabbed your way through it and fell off the other end.

This mod does a few things:
- The dialog itself is wrapped in `focus-trap-react`, which'll _very_ greedily grab onto focus and force it into the dialog, starting on the `X` button in the upper right.
- If the user has focused on a part of the dialog and hits the left/right arrow to go to the next tutorial, the focus will remain on whatever they had previously focused. If the focused item disappears (such as with "Teacher notes"), the focus will hang around in that area and the next tab will take it to the next available item.
- If the user clicked on a tutorial and then close the detail, the tutorial they had clicked on would gain focus. However, if the user changed tutorial via left/right arrow, focus would remain on the previously selected tutorial. Now when the detail is closed, the currently selected tutorial is focused and scrolled into view if necessary. It won't focus or scroll the selected tutorial until the dialog closes.

A few other points raised from this work:
- In some poking around, it looks like most of our dialogs don't properly behave in this manner and tabbing in them will lose focus, which is not ADA compliant. `focus-trap-react` is a wrapper around `focus-trap`, which we can use to add in this ability to our non-react based dialogs. This should give us a consistent experience until we get it refactored out.
- IMHO, the explore page would benefit from having the list of filters be static while the tutorials themselves scroll. As is, the user needs to scroll all the way back to the top.
- The page could also probably benefit from a refactor - it's just older style class based components, and more modern react might be easier to manage. I ended up deciding it wasn't urgent enough to push through atm.
- If you close the tutorial detail view by clicking outside of the modal, your focus will _not_ update, and in fact you will lose focus entirely.

## Links

jira ticket: [SL-592](https://codedotorg.atlassian.net/jira/software/c/projects/SL/issues/SL-592)

## Testing story

Basically, check all the scenarios listed above.
- Open tutorial detail, confirm you can tab through links w/o scrolling the page.
- Switch tutorials and confirm focus moves to the selected tutorial upon close.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/tutorialExplorer/tutorial.jsx [hunk 1]
```diff
@@ -39,6 +39,7 @@ export default class Tutorial extends React.Component {
         onKeyDown={this.keyboardSelectTutorial}
         tabIndex="0"
         role="button"
+        data-tutorial-code={this.props.item.code}
```
Comment (thomasoniii) [inline apps/src/tutorialExplorer/tutorial.jsx:42]: I don't have a feel for how bad this may be - could someone chime in?

This is leaking the tutorial's unique code value into the html of the page, making it more accessible to the user. They could potentially access in other ways (such as using the react dev tools), but this definitely makes it more visible since it's just in the source.

Is that something we're worried about? I'm assuming it's just a unique ID that's otherwise potentially visible so it's unimportant, but just in case it should be kept hidden, could somebody confirm?

And, assuming it should be switched, is there a different value that'd be better?

File: apps/src/tutorialExplorer/tutorialDetail.jsx [hunk 2]
```diff
@@ -200,7 +250,11 @@ export default class TutorialDetail extends React.Component {
                       rel="noopener noreferrer"
                       onClick={this.startTutorialClicked}
                     >
-                      <button type="button" style={{marginTop: 20}}>
+                      <button
+                        type="button"
+                        style={{marginTop: 20}}
+                        tabIndex={-1}
```
Comment (thomasoniii) [inline apps/src/tutorialExplorer/tutorialDetail.jsx]: This page has a `button` wrapped in an `a`, which isn't a great way to structure it. We should really either use a button or a link, but not both.

I opted to leave it for now for fear of breaking something structurally on the page, but I did give the inner button a negative tab index. Otherwise, the user would tab to the link, then tab to the button, and then onto the next one, which is confusing. This ensures only the link is tabbed.

Is there a reason why it's wrapped in this manner? And, if not, would we have a preference between using just the button or just the link? I'd lean towards the button with an `onClick` handler, but a stylized link would also be fine.

File: apps/src/tutorialExplorer/tutorialDetail.jsx [hunk 3]
```diff
@@ -42,6 +50,44 @@ export default class TutorialDetail extends React.Component {
...
+    /*
+      We've kept a ref to our dialog container, so when the user is adjusting focus we look to see if we're
+      focusing onto an element that's still in the modal. If so, then awesome! We're done. If not, then we look
+      in our modal and find the first button (which is the X to close) or the last link (which could vary based upon
+      what's in there).
+
+      If we're shifting focus away from the first button (and have left the modal), then we wrap around to the last link.
+      Otherwise, if we're leaving, we return to the start at the first button.
+
+      This feels very clunky. :-(
```
Comment (breville) [inline apps/src/tutorialExplorer/tutorialDetail.jsx]: Not a fan of shipping something we are disappointed about.  This UI is very high usage and we should be highly confident in our approach.

File: apps/src/tutorialExplorer/tutorialDetail.jsx [hunk 4]
```diff
@@ -42,6 +50,44 @@ export default class TutorialDetail extends React.Component {
...
+
+      But it's to ensure that the user can only tab within the modal while it is open and cannot shift focus outside.
+
+     */
+    if (this.containerRef.current) {
+      const inModal = this.containerRef.current.contains(event.target);
+      if (!inModal) {
+        const firstButton =
+          this.containerRef.current.getElementsByTagName('button')[0];
+        const links = this.containerRef.current.getElementsByTagName('a');
```
Comment (jmkulwik) [inline apps/src/tutorialExplorer/tutorialDetail.jsx]: I'm concerned about the fragility of this section. There are a number of reasons we might add a button or another tab-able element to the end of this dialog, which would break this functionality. This leans too far on the side of hard-coding a fix, and in this case, I am going to say this comment is blocking, change required.

Now that aside, this does bring up a deeper question of how much do we prioritize a fix vs future maintainability. In your PR description, you brought up a number of changes that you recommend, but are too big for this task. For example, you mentioned using a component library dialog that addresses this issue more holistically. Normally our approach in these cases is to improve things as much we can while also keeping the code and user experience both at good quality. Then make a task to follow up with a bigger, holistic fix.

In this case, we do have a [component library dialog](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/componentLibrary/StylizedBaseDialog.jsx) that would benefit from what you're describing. That leaves two choices, either of which I'm happy to support and make space for in sprint planning.
1. Slim this PR down so it only changes the code in ways you're happy about, even if that means the user experience isn't perfect yet. Then make a task that covers the work to get us the rest of the way there.
2. Make a pre-factor PR that adds correct tab navigation to the component library dialog. Then migrate this to use that dialog (those two pieces could probably happen in either order).

Either is a good option and there are likely other good options too. Happy to chat more too!

File: apps/src/tutorialExplorer/tutorialSet.jsx [hunk 5]
```diff
@@ -24,8 +24,30 @@ export default class TutorialSet extends React.Component {
   tutorialClicked = item =>
     this.setState({showingDetail: true, chosenItem: item});
 
-  tutorialDetailClosed = () =>
+  tutorialDetailClosed = () => {
+    // okay, so while the tutorialDetail window is open, it's possible that the focus has changed
+    // the user could've hit left/right arrows while the detail was open and moved focus, or also could've
```
Comment (jmkulwik) [inline apps/src/tutorialExplorer/tutorialSet.jsx]: Super tiny NIT: 
`okay` -> `Okay` and `has changed // the user` -> `has changed. // The user`
Just a norm here. The reason behind it: starting sentences with capital letters and ending with periods makes reading and understanding comments just a little easier.

File: apps/src/tutorialExplorer/tutorialSet.jsx [hunk 6]
```diff
@@ -24,8 +24,30 @@ export default class TutorialSet extends React.Component {
...
+    // item in the tutorial list and focus it, and also make sure to scroll it into view because it's possible
+    // the user navigated through enough of them to be off screen.
+    const tutorialDiv = document.querySelectorAll(
+      `[data-tutorial-code="${this.state.chosenItem.code}"]`
+    )[0];
+    if (
+      document.activeElement.hasAttribute('data-tutorial-code)') &&
+      document.activeElement !== tutorialDiv
+    ) {
+      // when in doubt, set timeout! _something_ is happening here that's fiddling with the focus, so we
```
Comment (jmkulwik) [inline apps/src/tutorialExplorer/tutorialSet.jsx]: `when in doubt, set timeout!` 
^ For my own learning, can you share some more details? This is my first time seeing setTimeout used in this way, and I'm curious to learn more (is there a link you can share for further reading)?

File: apps/src/tutorialExplorer/tutorialSet.jsx [hunk 7]
```diff
@@ -24,8 +24,30 @@ export default class TutorialSet extends React.Component {
...
+      // just defer it to the next run through of the even loop. if you remove this, the focus call will
+      // fail, although the scrollIntoView will work.
```
Comment (jmkulwik) [inline apps/src/tutorialExplorer/tutorialSet.jsx]: NIT: regarding "if you remove this, the focus call will fail..."
It's a small thing, but using the "personal you" in this warning comment has the effect of implicitly assigning blame, which can make the next developer slightly less likely to trust themselves to change the code and therefore the code slightly less likely to be maintained - things we generally want to avoid. Using the "[generic you](https://ell.stackexchange.com/questions/110889/is-using-you-to-refer-to-anyone-not-the-person-youre-talking-to-a-known-sp)" has less of this effect and 3rd person has even less, so those may work better in this case. For example, rewriting this as "If setTimeout is removed, the the focus call will fail..." removes the implicit blame while retaining the warning.

**Comment by breville** [comment]
There seems to be an issue in which, if the page is scrolled down, clicking on a tile scrolls back to the top rather than opens a dialog, and the dialog only shows when the page is scrolled up near the top.  Here is a repro after a fresh load on my localhost:

https://user-images.githubusercontent.com/2205926/231919398-b832b211-01c9-4259-909d-ed9d79f790b9.mov

**Review by breville — CHANGES_REQUESTED** [review]
Let's talk this through before proceeding.

File: apps/src/tutorialExplorer/tutorialDetail.jsx (hunk 3 above)
Comment (thomasoniii) [reply apps/src/tutorialExplorer/tutorialDetail.jsx]: Just because I said it was clunky didn't mean I'm not confident in the approach. 😄 

This is the accepted way to do it when rolling your own approach, it's just always annoying when you have to hop outside of react and directly deal with dom manipulations instead of handling it all in the component hierarchy itself. 

Doing it all inside of react would require substantial additional refactoring to all of the components in here, which I felt was better to defer atm.

File: apps/src/tutorialExplorer/tutorialSet.jsx (hunk 5 above)
Comment (thomasoniii) [reply apps/src/tutorialExplorer/tutorialSet.jsx]: done!

File: apps/src/tutorialExplorer/tutorialSet.jsx (hunk 6 above)
Comment (thomasoniii) [reply apps/src/tutorialExplorer/tutorialSet.jsx]: Somewhere during the `TutorialSet` re-render focus is being stolen. I'm not 100% sure where it happens, though I'd guess it has to do with the fact that the full tutorial list is re-rendered when the dialog is closed. I'm not sure.

`setTimeout` with a `0` just bumps the action through to the next available pass of the event loop. In this case it defers the focus until after the renders have completed to ensure we're working with a pristine and current dom and then does the focus selection at that point so we can be sure it doesn't get lost later.

Because the contents of the page aren't changing, we _technically_ could do the scroll outside of the `setTimeout`, but to me it makes more logical sense to focus and scroll at the same time, so I deferred 'em both.

That bit of wisdom `when in doubt, set timeout!` is an old joke from an old job - our codebase was riddled with race conditions, and it was annoyingly common to need to defer actions to the next tick in the event loop to make sure they happened in the right order.

For a complete react solution, I'd probably re-tool the page so that the tutorial list always keeps track of the selected item and highlights it independently of the browser focus, and also enhance the navigation so the user could use arrow keys to navigate (including up/down). If the list was always tracking whatever was selected, we could also defer focus into a `useEffect` hook, if this were refactored into a functional component.

**Comment by thomasoniii** [comment]
> There seems to be an issue in which, if the page is scrolled down, clicking on a tile scrolls back to the top rather than opens a dialog, and the dialog only shows when the page is scrolled up near the top. Here is a repro after a fresh load on my localhost:

Ooh. Good catch on that one. I thought I'd tested that, but must not have. 😠 

It was storing the starting scroll position on mount, but because the component is mounted once and kept forever, it got out of date. Re-tooled it so it behaves now. tyvm!

File: apps/src/tutorialExplorer/tutorialDetail.jsx (hunk 4 above)
Comment (thomasoniii) [reply apps/src/tutorialExplorer/tutorialDetail.jsx]: That's fair. I re-tooled it to use a more robust selector that'll catch all focusable elements on either end of the dialog.

 As a side note, I just added that selector text into this file, but it's a candidate to move to a separate library. OTOH, structurally we should largely discourage directly querying the dom in favor of staying in the react world, so keeping it in here to refactor out of existence later could be a better approach.

I would say that this PR is about as slimmed down as it can get to still properly address the issues, and that it is pretty surgical and self contained (it adds a new function, some additional set up (most of which is in self-contained new lifecycle methods), and the code to scroll element into view (which is also a single block).

As an alternative approach, the tiniest possible fix is to add an `autoFocus` attribute onto the `X` button of the dialog, but it doesn't _really_ work (the user could still tab off of the dialog (above or below), it steals focus from the tutorial list and never returns it, it doesn't address re-selection from navigation inside the dialog, etc). So this is the compromise slimmed down version to better address the request w/o causing more problems.

Swapping in a new dialog is a plausible option that'd help, but that alone wouldn't be enough to address the problems with changing tutorial focus from within the dialog via the area keys. And unfortunately that dialog component also has issues with focus falling off the dialog itself. So that dialog _also_ requires the same modifications added to the tutorial detail here. Not saying that it shouldn't be done there as well - just that the feature is scope creeping rapidly.

There are some other issues and inefficiencies in the tutorial page that we could address with a bigger scale refactor (such as the tutorial list constantly re-rendering), but again - scope creep way beyond the initial request.

Here's what I'd suggest:

1. Leave this PR (mostly?) as is (once everyone's confident in its behavior. I did address the noted ones already (thanks for the catches!)).
2. Re-tool the shared component dialog to ensure focus doesn't fall off of it (which is basically moving the `focusing` mods from here to that dialog), but maybe we should even skip over that because of note 3 here.
3. Then we can consider refactoring `tutorialDetail` to use the shared dialog, but honestly - we should also have a discussion about using a shared component library. `StylizedBasedDialog` appears to be mostly 2 years old and `BaseDialog` appears to be mostly 6 years old (looking at the git blames), and they're just simply more code we need to maintain. If it's viable to look at a third party system (material, blueprint, ant-design, whatever), then we could swap in something we don't need to maintain and that's known to already be ADA compliant. But I wouldn't want to just drop in a component system willy nilly, and if we do it, it should be with an eye towards migrating everything to it eventually, and with some sort of plan in place to do it.
4. Finally, we can look into a larger refactor of the tutorial explorer to more modern react, deal with the render inefficiencies, pin the filter list, keep closer track of selections, get rid of jQuery, etc.

Also happy to chat more and hopefully bring up your confidence in the proposed mods or discuss alternatives, as well as discuss what else we may want to do with the page going forward (if anything).

File: apps/src/tutorialExplorer/tutorialSet.jsx (hunk 6 above)
Comment (thomasoniii) [reply apps/src/tutorialExplorer/tutorialSet.jsx]: Ooh! Even better. It was the `this.setState({showingDetail: false, chosenItem: null});` call that was stealing focus, and by moving that ahead of the focus/scroll combo it didn't steal it away and we could just do the focus/scroll synchronously w/o deferring to the next pass. So I dropped it.

**Comment by breville** [comment]
Sorry for the delay on this.  Thanks for digging into the interesting world of the "focus trap"!  I'd not heard of this before, and it indeed gets quite tricky.  I got my hands dirty in parallel, just to try and understand this space a bit better.

### Alternative tech

I found a couple interesting posts, with a novel approach at https://stackoverflow.com/a/28635872, and a few other options at https://stackoverflow.com/a/50179843.

I tried the `focus-trap-react` suggestion (https://github.com/focus-trap/focus-trap-react), and interestingly, just wrapping the modal in a `<FocusTrap>` was enough to get most of the core functionality inside the dialog.  (See [here](https://github.com/code-dot-org/code-dot-org/pull/51365/commits/544fc7c1f45c9477fef7b606cb8797103507d843?w=1).)

Does it make sense to use a third-party library here, since it seems pretty comprehensive and supported?  Or is there an argument to roll our own?  

And if there is, could we use a similar model so that we can wrap our other dialog implementations with a common tag?  It does feel like the right thing is to separate this logic from a particular dialog implementation.  We have some precedent for this kind of approach in Tutorial Explorer with the [Image](https://github.com/code-dot-org/code-dot-org/blob/2a25deb24eb35b0a7d28db39d29789702ef8a2be/apps/src/tutorialExplorer/image.jsx) class. 

### Cross-browser issues

We support a fairly wide range of browsers.  I tried three (Chrome, Safari, on Firefox) on macOS but we should also try the remainder, since focus seems to have OS- as well as browser-specific peculiarities.

Of those three, Chrome mostly works, Safari works apart from skipping the image, and Firefox only finds a couple things to focus upon.  It turns out that Firefox needs a macOS setting -- System Settings | Keyboard | Keyboard navigation to "on" -- before tab navigation works more as expected.  (The makers of `focus-trap-react` have accepted this.  If we roll our own, is this something we also need turned on?).   With the setting on, Firefox works better, though with one extraneous focus step.

Separately, an aesthetic issue is that each browser does its focus visualisations differently, and not all of them look very good.  It's tempting to draw our own consistent focus visuals, but it seems there is some advocacy for not touching these defaults so that each browser can continue to iterate on its own accessibility UI.  

### Styling

I had a play with the styling (using `focus-trap-react`) and prepped some changes that might prove useful, so that focus highlights look nicer, and don't get cut off, etc.  This also includes an improvement to the button issue noted separately, fixes to the incorrect focus steps in Safari and Firefox, and some general appearance improvements along the way.  (See [here](https://github.com/code-dot-org/code-dot-org/pull/51365/commits/e2bcd2f4aa6244629d6eb320c290054686f19888?w=1).)

Here are videos of the three browsers successfully cycling focus with these visual improvements: 

#### Chrome
https://user-images.githubusercontent.com/2205926/234547688-b6646bc5-677b-4932-a1c1-06c1aa9e43f9.mov

#### Safari
https://user-images.githubusercontent.com/2205926/234547749-a9649e50-08c2-4ffb-86ab-613987a1e83e.mov

#### Firefox
https://user-images.githubusercontent.com/2205926/234547798-c8e0b4bc-c40c-465e-b56b-e61309c84223.mov

Feel free to leverage this styling work!

### Other features

The scroll-jump to the last-tutorial viewed is a great idea; thanks for doing that.

File: apps/src/tutorialExplorer/tutorialDetail.jsx (hunk 2 above)
Comment (breville) [reply apps/src/tutorialExplorer/tutorialDetail.jsx]: > Is there a reason why it's wrapped in this manner? And, if not, would we have a preference between using just the button or just the link? I'd lean towards the button with an `onClick` handler, but a stylized link would also be fine.

I don't think it needs to stay this way, and was just done so that we'd have something that acted like a link but looked like a button.  I think it's best to leave the `<a>` intact so that screen-readers and browsers themselves get a proper preview of the target URL, but to just have a `<div>` styled to look like a button.  I prepped such a fix in the same [diff](https://github.com/code-dot-org/code-dot-org/pull/51365/commits/e2bcd2f4aa6244629d6eb320c290054686f19888?w=1) mentioned elsewhere.

File: apps/src/tutorialExplorer/tutorialSet.jsx [hunk 8]
```diff
@@ -24,8 +35,27 @@ export default class TutorialSet extends React.Component {
...
+    this.setState({showingDetail: false, chosenItem: null}, () => {
+      if (document.activeElement !== tutorialDiv) {
+        setTimeout(() => {
+          tutorialDiv.focus();
+          if (!isInViewport(tutorialDiv)) {
+            tutorialDiv.scrollIntoView({behavior: 'smooth', block: 'nearest'});
+          }
+        }, 0);
+      }
+    });
```
Comment (thomasoniii) [inline apps/src/tutorialExplorer/tutorialSet.jsx:57]: There is a timing issue here, which seems to be related to how tightly `focus-trap-react` is grabbing focus and exactly when it's releasing it.

Best option I was able to find was to fall into the `setState` callback (deferring until state is properly fully updated) and _still_ needing a timeout to defer to the next pass of the event loop.

**Comment by thomasoniii** [comment]
Updated the PR to merge in the mods from https://github.com/code-dot-org/code-dot-org/pull/51365, including using `focus-trap-react` + the styling, and then some other mods that fell out of it regarding updating retargeting the focus upon close.

Hopefully gtg at this point.

File: apps/src/tutorialExplorer/tutorial.jsx (hunk 1 above)
Comment (breville) [reply apps/src/tutorialExplorer/tutorial.jsx:42]: This should be fine.

**Review by breville — APPROVED** [review]
Thanks for these improvements!  Looking great, both from reviewing code and testing the latest.

Would probably be worth double-checking everything looks good on a variety of Windows & mobile browsers, though given the scoping of this change, we can probably do this after it's gone live if easier.

**Comment by megcrenshaw** [comment]
Hey folks. I used this PR and playing around on https://code.org/learn to help me build an accessible dialog!

One thing I noticed: When I use the screenreader tools to navigate (on Mac VoiceOver, it's control+option+arrow keys), if I navigate away from the modal by pressing control+option+up, it's very hard to find the modal again, and I instead find myself navigating around the "background" of the page. Demo:

https://github.com/code-dot-org/code-dot-org/assets/9142121/1c9c74a3-4de4-4e22-b264-228a66851292

My first instinct was, "Yikes, FocusTrap doesn't work!" But then I played around with adding `role="dialog"` to the modal itself, and it seems to have fixed the problem. I believe you all could use the same strategy here? If anyone tries it and it works, let me know!

**Comment by breville** [comment]
> I believe you all could use the same strategy here? If anyone tries it and it works, let me know!

@megcrenshaw Thanks so much for this information.  I tried it out by adding `role=dialog` separately to a couple different divs -- one with class `modal-dialog` and the other with class `modal-content` -- and unfortunately when using VoiceOver I was able to escape out of the modal by pressing `ctrl-option-[left arrow]`.  It took us from the modal's `X` over to the filter options on the backing page.  This was on macOS Chrome and in at least one case double-checked on macOS Safari.

It sounds like this is a known behaviour when only using `role="dialog"`: https://stackoverflow.com/questions/53561764/trap-focus-with-in-popup-modal-only-ios-voiceover

It suggests using `aria-hidden="true"`, but it sounds like there are also nuances with that approach which wouldn't be super easy for us without some bigger changes to our DOM structure: https://stackoverflow.com/questions/21828152/nested-aria-hidden

**Comment by breville** [comment]
> Thanks so much for this information. 

@megcrenshaw After seeing how things work at https://studio.code.org/catalog, it looks like we need `aria-modal="true" role="dialog"` and then VoiceOver navigation does appear to be captured!
